vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs

```diff
@@ -1674,114 +1672,167 @@ namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
+                    projNode.UpdateMSBuildState()
+                    projNode.SetProjectFileDirty(projNode.IsProjectFileDirty)
+                    projNode.ComputeSourcesAndFlags()
                     VSConstants.S_OK
 
             interface IVsUpdateSolutionEvents3 with
+
                 member x.OnBeforeActiveSolutionCfgChange(_oldCfg, _newCfg) =
+
                     // this will be called for each project, but wait dialogs cannot 'stack'
                     // i.e. if a wait dialog is already open, subsequent calls to StartWaitDialog
                     // will not override the current open dialog
-                    waitDialog <-
-                        {
-                            WaitCaption = FSharpSR.GetString FSharpSR.ProductName
-                            WaitMessage = FSharpSR.GetString FSharpSR.UpdatingSolutionConfiguration
-                            ProgressText = None
-                            StatusBmpAnim = null
-                            StatusBarText = None
-                            DelayToShowDialogSecs = 1
-                            IsCancelable = false
-                            ShowMarqueeProgress = true
-                        }
-                        |> WaitDialog.start projNode.Site
-                        |> Some
+                    if waitDialog.IsNone  then 
+                        waitDialog <-
+                            {
+                                WaitCaption = FSharpSR.GetString FSharpSR.ProductName
+                                WaitMessage = FSharpSR.GetString FSharpSR.UpdatingSolutionConfiguration
+                                ProgressText = None
+                                StatusBmpAnim = null
+                                StatusBarText = None
+                                DelayToShowDialogSecs = 1
+                                IsCancelable = false
+                                ShowMarqueeProgress = true
+                            }
+                            |> WaitDialog.start projNode.Site
+                            |> Some
 
                     VSConstants.S_OK
 
                 member x.OnAfterActiveSolutionCfgChange(_oldCfg, _newCfg) =
+
+                    // We have updated the MSBuild state of each project.  Now we can call ComputeSourcesAndFlags
+                    projNode.ComputeSourcesAndFlags()
                     match waitDialog with
                     | Some x ->
                         x.Dispose()
                         waitDialog <- None
-                    | None -> ()
+                    | _ -> ()
                     VSConstants.S_OK
               
             interface IVsUpdateSolutionEvents4 with
                 member x.OnActiveProjectCfgChangeBatchBegin() =
-                    batchState <- BatchWaiting
+                    ()
+
                 member x.OnActiveProjectCfgChangeBatchEnd() =
-                    batchState <- NonBatch
+                    projNode.UpdateMSBuildState()
```

Comment:
All good questions.  TBH I just fiddled around for about 3 days (yes really) until it seemed to work in every scenario I tried it in.  Almost every other variation seemed to fail in some way or another.

I didn't however try a batch build.  I had misunderstood what that actually was, sorry.

You are correct that the dialog goes away early though another "computing sources and flags for project" replaces it.  

I'll see if I can find a less invasive set of changes.  Perhaps there is just one fix needed in this code after all.